Add initial Python unit/performance tests for TIFF loader module

gigony opened the pull request:
The main change is for addressing parts of #38 and #61.

- Add initial unit/perf test cases for TIFF loader
- Add license texts for python tests

This also has the following minor changes:
- Fix some typo in the doc.
- Remove cupy_rc channel in README.md
- Update name in Github templates
- Update calculation of patch_down_count

And build/doc related changes
- Update build script and contributing guide
- Fix broken links in docs
- Sort import statements by using isort

```bash
./run test                             # execute all tests
./run test python                      # execute all python tests
./run test python unit                 # execute all python unit tests
./run test python unit skimage         # execute all python unit tests in `skimage` module
./run test python unit clara           # execute all python unit tests in `clara` module
./run test python performance          # execute all python performance tests
```


Current test folder structure looks like below:

```bash
python/cucim/tests/
├── __init__.py
├── conftest.py
├── fixtures
│   └── testimage.py
├── integration
│   ├── __init__.py
│   └── clara
│       └── __init__.py
├── performance
│   ├── __init__.py
│   └── clara
│       ├── __init__.py
│       └── test_read_region_memory_usage.py
├── test_cucim.py
├── unit
│   ├── __init__.py
│   └── clara
│       ├── __init__.py
│       ├── test_image_cache.py
│       ├── test_load_image.py
│       ├── test_load_image_metadata.py
│       └── test_tiff_read_region.py
└── util
    ├── gen_image.py
    └── gen_tiff.py
```

<!--

Thank you for contributing to ___PROJECT___ :)

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

4. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

5. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present) and replace
   it with `[REVIEW]`. If assistance is required to complete the functionality,
   for example when the C/C++ code of a feature is complete but Python bindings
   are still required, then add the label `[HELP-REQ]` so that others can triage
   and assist. The additional changes then can be implemented on top of the
   same PR. If the assistance is done by members of the rapidsAI team, then no
   additional actions are required by the creator of the original PR for this,
   otherwise the original author of the PR needs to give permission to the
   person(s) assisting to commit to their personal fork of the project. If that
   doesn't happen then a new PR based on the code of the original PR can be
   opened by the person assisting, which then will be the PR that will be
   merged.

6. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please do not
   rebase your branch on main/force push/rewrite history, doing any of these
   causes the context of any comments made by reviewers to be lost. If
   conflicts occur against main they should be resolved by merging main
   into the branch used for making the pull request.

Many thanks in advance for your cooperation!

-->

jakirkham commented on ci/gpu/build.sh:
```diff
@@ -52,7 +52,9 @@ gpuci_logger "Activate conda env"
 
 gpuci_logger "Install dependencies"
-gpuci_conda_retry install -y -c rapidsai-nightly \
+conda conda list --revisions
+
+gpuci_conda_retry install -y -c conda-forge -c rapidsai-nightly \
```
```suggestion
gpuci_conda_retry install -y -c rapidsai-nightly \
```

jakirkham commented on ci/gpu/build.sh:
```diff
@@ -82,8 +84,14 @@ gpuci_conda_retry build -c ${LIBCUCIM_BLD_PATH} -c conda-forge -c rapidsai-night
 gpuci_logger "Installing cuCIM and its dependencies"
-gpuci_conda_retry install -y -c ${LIBCUCIM_BLD_PATH} -c ${CUCIM_BLD_PATH} -c rapidsai-nightly \
-    "rapids-build-env=$MINOR_VERSION.*" \
+
+conda conda list --revisions
+
+gpuci_conda_retry install -y \
+    --override-channels \
+    -c ${LIBCUCIM_BLD_PATH} \
+    -c ${CUCIM_BLD_PATH} \
+    -c conda-forge \
```
```suggestion
gpuci_conda_retry install -y -c ${LIBCUCIM_BLD_PATH} -c ${CUCIM_BLD_PATH} -c rapidsai-nightly \
     "rapids-build-env=$MINOR_VERSION.*" \
```

jakirkham commented on ci/gpu/build.sh:
```diff
@@ -109,6 +117,5 @@ else
     cat /etc/hosts
 
     gpuci_logger "Python py.test for cuCIM"
-    cd $WORKSPACE/python/cucim
-    py.test --cache-clear -v --ignore-glob . --rootdir=src
+    ./run test python all
```
Is this still needed?

gigony replied:
Yes, it is still needed. I have integrated running test cases from both skimage and clara through `run` script.

jakirkham reviewed (commented):
Thanks Gigon! 😄

Had a couple suggestions about CI changes that I think we can now drop as well as one question

ajschmidt8 reviewed (approved):
Approving `ops-codeowner` file changes

gigony commented:
@jakirkham now it is building. Could you please help review this?

grlee77 commented on python/cucim/tests/unit/clara/test_image_cache.py:
```diff
@@ -0,0 +1,216 @@
+    # same with `img.resolutions["level_dimensions"][0]`
+    image_size = img.size('XY')  # 32x24
+    tile_size = img.resolutions['level_tile_sizes'][0]  # 16x16
+    patch_size = (tile_size[0] * 2, tile_size[0] * 2)
```
```suggestion
    patch_size = (tile_size[0] * 2, tile_size[1] * 2)
```

tile_size is a 2-tuple?

grlee77 commented on python/cucim/tests/unit/clara/test_tiff_read_region.py:
```diff
@@ -0,0 +1,157 @@
+        cucim_arr = np.asarray(cucim_img.read_region(start_pos, size))
+
+        # Not channel values are zero, so we need to check that.
+        channel_value_count = np.apply_along_axis(np.count_nonzero, 2,
+                                                  cucim_arr)
+        count_all_zero = np.count_nonzero(channel_value_count == 0)
+        assert count_all_zero == 0
```
```suggestion
        channel_value_count = np.count_nonzero(cucim_arr, axis=2)
        assert np.all(channel_value_count > 0)
```
should be equivalent (avoid `apply_along_axis` which is usually pretty inefficient)

grlee77 commented on python/cucim/tests/unit/clara/test_tiff_read_region.py:
```diff
@@ -0,0 +1,157 @@
+    for (start_pos, size) in region_list:
+        cucim_arr = np.asarray(cucim_img.read_region(start_pos, size))
+        # Not channel values are zero, so we need to check that.
+        channel_value_count = np.apply_along_axis(np.count_nonzero, 2,
+                                                  cucim_arr)
```
```suggestion
        channel_value_count = np.count_nonzero(cucim_arr, axis=2)
```

grlee77 commented on python/cucim/tests/unit/clara/test_tiff_read_region.py:
```diff
@@ -0,0 +1,157 @@
+    for (start_pos, size) in region_list:
+        cucim_arr = np.asarray(cucim_img.read_region(start_pos, size))
+        # All channel values should be zero, so we need to check that.
+        channel_value_count = np.apply_along_axis(np.count_nonzero, 2,
+                                                  cucim_arr)
```
```suggestion
        channel_value_count = np.count_nonzero(cucim_arr, axis=2)
```

grlee77 commented on python/cucim/tests/unit/clara/test_tiff_read_region.py:
```diff
@@ -0,0 +1,157 @@
+    for level in range(level_count):
+        cucim_arr = np.asarray(cucim_img.read_region(start_pos, size, level))
+        # Not channel values are zero, so we need to check that.
+        channel_value_count = np.apply_along_axis(np.count_nonzero, 2,
+                                                  cucim_arr)
```
```suggestion
        channel_value_count = np.count_nonzero(cucim_arr, axis=2)
```

grlee77 commented on python/cucim/tests/unit/clara/test_tiff_read_region.py:
```diff
@@ -0,0 +1,157 @@
+    assert array_interface['shape'] == tuple(whole_img.shape)
+    assert array_interface['version'] == 3
+
+    def test_cuda_array_interface_support(testimg_tiff_stripe_32x24_16_jpeg):
```
Doesn't this whole function need to be un-indented? why is it nested within `test_array_interface_support`?

grlee77 commented on run:
```diff
@@ -656,6 +656,127 @@ copy_data_() {
+
+install_python_test_deps_() {
+    # Package name has '<package name>[:<conda package name>]' format.
+    check_install_python_dep_ pytest pytest-cov pytest-lazy-fixture psutil GPUtil tifffile imagecodecs openslide-python
```
may want to keep these in a separate `requirements-test.txt` file or similar? If this is the only place they are used, then it is probably not a big deal either way.

gigony replied:
That's good idea! Will also use fixed versions of the packages.

grlee77 reviewed (changes requested):
Hi Gigon, this looks mostly good to me. I left just some minor comments.

The style of passing the image parameters as a string with `:` separator seems a little un-Pythonic, but this is only for test code, not part of the public API, so I don't have a problem with it.

I have not yet tried running this code, but can try that on the 5th or 6th.

gigony reviewed (commented):
Thank you @grlee77 for the all valuable review! Please help review the updated code.

grlee77 reviewed (approved):
Thanks, the test dependency installation looks much cleaner now in the `run` script.

It would be nice to also update the developer docs to mention how the tests should be run (in CONTRIBUTING.md?)

gigony commented:
@grlee77 Addressed your comments.
Please help review them.

The following commits are added:
- [Update build script and contributing guide](https://github.com/rapidsai/cucim/pull/62/commits/866b6071a19eb6590511702aa2770efc7b9cbe9c)
- [Fix broken links in docs](https://github.com/rapidsai/cucim/pull/62/commits/2b75648917b0151028b7429edd7e3997e99e5b0c)
- [Sort import statements by using isort](https://github.com/rapidsai/cucim/pull/62/commits/86535deb468f2c614945b304dc4acf4a6f0db0fb)

grlee77 commented on python/cucim/ci/bootstrap.py:
```diff
@@ -1,16 +1,11 @@
-from __future__ import absolute_import
-from __future__ import print_function
-from __future__ import unicode_literals
+from __future__ import absolute_import, print_function, unicode_literals
```
We shouldn't need the `__future__` imports here since we are Python 3.x only

gigony replied:
Thanks Greg!
oh, the file is not used in RAPIDS CI/CD but will remove it.

grlee77 commented on python/cucim/setup.py:
```diff
@@ -1,17 +1,13 @@
 # -*- encoding: utf-8 -*-
-from __future__ import absolute_import
-from __future__ import print_function
+from __future__ import absolute_import, print_function
```
We shouldn't need the `__future__` imports here since we are Python 3.x only

gigony replied:
The file is not used in RAPIDS CI/CD but will remove it.

grlee77 commented on CONTRIBUTING.md:
```diff
@@ -96,56 +96,190 @@ CUDA/GPU requirements:
+# Building and Testing cuCIM from Source
 
-### Build from Source
+First, please clone the cuCIM's repository
```
```suggestion
First, please clone cuCIM's repository
```

grlee77 commented on CONTRIBUTING.md:
```diff
@@ -96,56 +96,190 @@ CUDA/GPU requirements:
+
+Conda can be used to setup GCC 9.x, CUDA Toolkit (including nvcc) 11.0.x, and other dependent libraries (as shown in `./conda/environments/env.yml`) for building cuCIM.
+
+Otherwise, you may need to install (such as zlib, xz, yasm) through your OS's package manager (`apt`, `yum`, and so on).
```
```suggestion
Otherwise, you may need to install dependencies (such as zlib, xz, yasm) through your OS's package manager (`apt`, `yum`, and so on).
```

grlee77 reviewed (approved):
I reviewed the recent commits, making only a few small suggestions

Thank you for adding the `clean` and `clean_cache` options to the run script!

jakirkham commented on conda/environments/env.yml:
```diff
@@ -16,8 +16,12 @@ dependencies:
   - scipy
   - python=3.8
   - cudatoolkit=11.0
+  - cudatoolkit-dev=11.0
```
Why is this needed?

gigony replied:
In Conda package, cudatoolkit package doesn't include `nvcc` binary, but nvcc is needed to build the project so added it. Does it make sense? Any other solution to do that? :)

jakirkham replied:
Yeah it doesn't. We are working on a longer term solution ( https://github.com/conda-forge/cudatoolkit-feedstock/issues/62 )

However this package was added for historical reasons to help users run the CUDA installer. I don't think we should be using it

gigony replied:
@jakirkham for now, what would be a good solution until the long-term solution is available?
Currently, without installing cudatoolkit-dev, it would try to find system's nvcc if available which cannot be isolated in the environment. Do you think we can keep the change(`cudatoolkit-dev`) until the long-term solution is available?

jakirkham replied:
Is it ok if we just rely on the system one for now? Or are there other issues with that?

Should add that others may just be using our prebuilt packages. So think this is more a question of what works for us

gigony replied:
> Is it ok if we just rely on the system one for now? Or are there other issues with that?
I thought `env.yml` is for setting up 'development environment' to build cuCIM locally from the source code and it is not for the runtime environment (it already has more than needed for just running cuCIM. e.g., yasm/compilers/cmake/flake8 etc).
 
I don't think all users (developer who want to contribute here) have the same nvcc version needed (in this case, CUDA 11.0) in the system so I think `cudatoolkit-dev` with the same version of `cudatoolkit` is needed.

May be we need to separate out `env.yml` -- one for dev, one for runtime.

jakirkham commented on conda/environments/env.yml:
```diff
@@ -16,8 +16,12 @@ dependencies:
   - python=3.8
   - cudatoolkit=11.0
+  - cudatoolkit-dev=11.0
+  - gxx_linux-64=9
```
Should we add the `cxx-compiler` package? This will handle the activation of the C++ compiler as well

Do we need the `c-compiler` too?

gigony replied:
If I remember correctly, Installing `- gxx_linux-64=9` also installed gcc_linux-64=9.

I am not sure how cxx-compiler/c-compiler package works. would it be better to use it?

jakirkham replied:
Yeah switching to `cxx-compiler` & `c-compiler` would preferable. They operate the same way as when Conda-Build would use them

gigony replied:
Updated to use cxx-compiler and c-compiler that install `gxx_linux-64=9`.
Thanks @jakirkham!

jakirkham replied:
Thanks Gigon :)

jakirkham reviewed (commented):
Thanks Gigon! 😄

This looks pretty good. Also like the addition of the `clean` command 🙂

Had a couple minor questions about the environment file

gigony commented:
Hi @jakirkham, can this PR be merged or do I need to address others?

jakirkham commented:
@gpucibot merge

jakirkham commented:
Thanks Gigon! 😄
